src/p2p/behaviour.rs

```diff
@@ -427,20 +427,14 @@ impl<Types: IpfsTypes> Behaviour<Types> {
         self.bitswap.want_block(cid, 1);
     }
 
-    // FIXME: it would probably be best if this could return a SubscriptionFuture, so
-    // that the put_block operation truly finishes only when the block is already being
-    // provided; it is, however, pretty tricky in terms of internal communication between
-    // Ipfs and IpfsFuture objects - it would currently require some extra back-and-forth
-    pub fn provide_block(&mut self, cid: Cid) {
+    pub fn provide_block(
+        &mut self,
+        cid: Cid,
+    ) -> Result<SubscriptionFuture<(), String>, anyhow::Error> {
         let key = cid.to_bytes();
         match self.kademlia.start_providing(key.into()) {
```

Comment:
I did not previously notice that start_providing was already being called. I don't think we should be doing that at all at the moment given the "client dht" vs. "server dht" completion.

Or how do you see the risk of putting a provider record in without having peers who know how to connect to us (and also knowing that we can be connected to)? I think this was the reason why go-ipfs 0.5 introduced the big DHT changes.

To reiterate, I am worried we might be poisoning the dht *unless* we are sure we can be connected to and are making sure our peerid => multiaddr[] record is in place.

Reply:
Will those records not be readily available when the other peers connect to us? Kademlia should be re-publishing them periodically. Other than that I don't have any strong feelings towards this, I can comment it out so that it can be reintroduced at a later stage.

Reply:
I don't know but I assume the publishing or the record putting is an act of finding all of the closest peers to key, and asking them to store it until expiration. Continuing on what I expect, we should only be providing these records in the event of us being one of the K closest peers to the key, but I guess there could be an optimization.

Probably best to do something like `if false { ... }` around this, with a comment. Is the key correct? I would had assumed that multihash was already enonugh? Perhaps we've only tried it v0 cids which have some compatible byte format?

Reply:
Fair enough, I've disabled that functionality for now.

As for the `Cid` working as a key, it all depends on how the other implementations are doing it I guess; if this is pure `Multihash` then of course we should change it; I'll do it in a follow-up if that's the case, as the `want_block` part will need to be adjusted too.

Reply:
This looks good.